sparkmagic/proposal.md

```diff
@@ -0,0 +1,49 @@
+# Proposal for Incubation
+
+## Subproject name
+
+sparkmagic
+
+## Development team and Advocate
+
+Subproject development team:
+
+* Alejandro Guerrero Gonzalez, Microsoft (`@aggFTW`)
+* Joel Zambrano, Microsoft (`@djoelz`)
+* Auberon Lopez, Cal Poly (`@alope107`)
+
+Steering Council Advocate
+
+* Brian Granger (`@ellisonbg`)
+
+## Subproject goals, scope and functionality
+
+Goals:
+* Make Spark usage from the IPython notebook as frictionless as possible by decoupling Spark code execution from Spark installation.
+* These magics will be packaged so that they can be pip-installable or conda-installable.
+* Allow any Jupyter installation to point to different Spark installations.
+
+Scope:
+* Spark will be the first back end provided by these magics, but it could easily be extended to other big data back ends.
+* Remote Spark execution through a REST endpoint which allows for Scala, Python, and R support as of Sept 2015.
+
+Functionality:
+* Enable users to remotely connect to a REST Spark back end.
+* Return of pandas dataframes from Spark when appropriate. This will allow users to integrate spark with the Python libraries available for pandas dataframes.
+
+## Audience
+
+Spark users who want to have the best experience possible from a Notebook. These users care about the libraries ecosystem the Jupyter community has created.
+
+## Other options
+
+Alternatives we know of:
+
+* Combination of IPython, R kernel (or rpy2), and Scala kernel for an in-Spark-cluster Jupyter installation. This does not allow the user to point to different Spark clusters. It might also result in resource contention (CPU or memory) between the Jupyter installation and Spark.
+* IBM's Spark kernel does not provide a REST endpoint, requires the installation of Jupyter in the cluster, and does not create pandas dataframes.
```

Comment:
> IBM's Spark kernel does not provide a REST endpoint, requires the installation of Jupyter in the cluster, and does not create pandas dataframes.

Personally, I've always treated the "IBM Spark Kernel" as a Scala kernel for Jupyter that has all the pieces needed to talk to Spark from Scala, and gives you SparkContext by default when you run it. No more, no less. It's not really all that different from the other Jupyter language kernels IMHO. 

>  I think talking to spark over REST makes tons of sense and is a huge step forward.

It is a step forward from having to run the entire Jupyter notebook server plus kernels on the same L3 network as the Spark workers. But keep in mind a pure REST API will likely not work well with Spark Streaming which is one of the huge draws of Spark.

>  It also enables a much more flexible set of ways for integrating with Jupyter as kernels no longer have to be run directly on the spark nodes.

Again, while a REST API is a nice simple first step toward remote access to a Spark cluster, there are potential benefits to keeping the kernels on / near the compute cluster but running the Notebook web server app remotely, namely:

* Compute driver (kernel) stays close to compute workers (Spark) stays close to (big) data
* You get streaming "for free" thanks to the Jupyter protocol and implementation (0mq + websockets)
* Enabling kernels to run remote from the Notebook server has other benefits beyond remote access to Spark 

The last is the impetus for the potential [Kernel Provisioner and Gateway API proposal](https://jupyter.hackpad.com/Jupyter-Kernel-Service-API-sZx2qqNHnY1). I've spent most my time so far documenting potential uses case for it, the third of which on that page could cover remote access to a Spark (or any) compute cluster. I'm still figuring out how to bring it forward (JEP? incubator?) and consulting with @rgbkrk, @freeman-lab, et al about how it fits with other efforts under way. That said, I think it will take some time to realize and so please don't see it as discouraging this magics + REST API proposal. I don't believe there needs to be one and only one way to get Jupyter to work with Spark.

/cc @lbustelo